keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,266 @@
+2. `NoExecuteTaintManager` watches for node and pod updates,  and performs `NoExecute` taint based pod eviction
+
+The existing  kube-controller-manager flag `--controllers` can be used to optionally disable the `NoExecuteTaintManager`.
+
+<p align="center"><image src="new-controllers.png" width=60%></p>
+
+### Implementation
+
+A new `NodeLifecycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.
+
+A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.   
+
+The creation and startup of the default taint-manager is performed by `kube-controller-manager`. A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or  roll back to the old `NodeLifecycleController`.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.
+
+##### Prerequisite testing updates
+
+Kubernetes already has a good coverage of node `No-Execute` eviction. We will add tests only for the changed code.
+
+#####  Unit tests
+- `pkg/controller/taintmanager`: 2023-05-23 - 0%  
+  - Enable and disable the new `TaintManager`
+  - The new `TaintManager` acts on node taints properly
+- `pkg/controller/apis/config/v1alpha1`: 2023-05-23 - 0% 
+  - Use the split `TaintManager` or the old combined controller
+
+#####  Integration tests
+ No integration tests are required. This feature is better tested with e2e tests.  
+
+##### E2E tests
+
+- Verify passing the existing E2E and conformance tests using the split `TaintManager`.
+- Verify the ability to disable the default `TaintManager`.
+
+### Graduation Criteria
+
+#### Alpha
+
+Since there are no new API changes, we can skip Alpha and go to Beta directly.
+
+#### Beta
+
+* Support `kube-controller-manager` flag `--controllers=-taint-manager` to disable the default taint-manager.
+* Unit and e2e tests completed and passed.
+* Performance and scalability tests to verify there is non-negligible performance degradation in node taint-based eviction.
+* Update documents to reflect the changes.
+
+#### GA
+
+* Fix all reported bugs if any.
+
+### Upgrade / Downgrade Strategy
+
+We’ll need to upgrade `kube-controller-manager` and` node-lifecycle-controller` and add a new controller `taint-manager`. Accordingly, a downgrade will revert to the previous version of `kube-controller-manager` and `node-lifecycle-controller` and remove the new added `taint-manager` . There should be no impact on the other components during upgrades and downgrades since user-exposed behavior should stay the same except a new flag to enable/disable the default `taint-manager`. This change should be clearly communicated to users.
```

Comment:
This doesn't make sense. kube-controller-manager is one process, which embeds several controllers (node-lifecycle-controller and taint-manager, when added). The upgrade and downgrade description here should point out that this only affects kcm, so there are no expected problems during upgrade and downgrade. 
Unless you start disabling the built-in taint-manager and using a custom built one, in which case you're approach will be different, but that case is again non-goal, so I'm not sure it's worth describing it here.

Reply:
I think we can reword in the perspective of user's angle:

For a user that doesn't use this feature, no matter which kind of upgrade/downgrade path is performed, the eventual result should be that both "applying NoExecute taints on nodes upon conditions" and "evict pods upon NoExecute taints" function well. (they don't care about the underlying code restructured or not)

For a user that uses this feature, downgrade would mean (in-tree) taint-manager will be enabled, no matter it's previously enabled or not; upgrade would mean (in-tree) taint-manager will be enabled, but can be disabled. We don't need to mention out-of-tree taint-manager, as 1) that's no different from an external controller, and 2) nothing can stop a customer runs duplicated or faulty taint-manager.

Reply:
How about the following

A feature gate `SeparateTaintManager`  enables and disables the new feature. When the feature is turned on, a user can use `kube-controller-manager`'s flag to disable the default `TaintManager`.

Reply:
I think the point is that what you are saying is part of the Design, or the question "how can the user disable this feature".
But it's not part of the upgrade/downgrade strategy.
The answer is "N/A" for this KEP